doc/mgmt/SONiC_PIM_SM_HLD.md

```diff
@@ -0,0 +1,1036 @@
+
+    1.1.(b) When the RP or an intermediate PIM router receives a PIM (\*,G) Join message, the message is passed to the Linux kernel and kernel delivers the PIM Join message and its associated payload to Pimd process.  PIM queries Zebra for RPF validation and then creates a (\*,G) entry with the RPF interface as its IIF. In case of RP, the IIF interface is the interface on which the RP address is configured. PIM queries IGMP database (inside Pimd) and updates the (\*,G) entry's OIF list. To complete the registration process RP needs the data packet in Pimd, so to get the multicast data in Pimd the pimreg interface is updated in the OIF list.  <br> <br>
+    
+2. After PIM creates the (\*,G) entry with the required information, PIM updates Linux kernel using a socket and kernel creates a mroute entry. <br> <br>
+   
+3. ipmcFpmSyncd registers with Linux kernel for RT netlink group RTNLGRP_IPV4_MROUTE and listens for RTM_NEWROUTE and RTM_DELROUTE RT netlink events.  When Pimd creates/deletes a multicast route entry, Linux kernel notifies ipmcFpmSyncd about multicast route entry creation/deletion using RTM_NEWROUTE and RTM_DELROUTE events. <br> <br>
+
+4. ipmcFpmsyncd processes the netlink message and updates the (*,G) route entry information into IPMC_ROUTE_TABLE in APPL_DB. <br> <br>
+
+5. ipmcOrch being an subscriber for IPMC_ROUTE_TABLE in APPL_DB , it receives the multicast route entry information  updated by ipmcFpmsyncd. <br> <br>
+
+6. After processing the received information, ipmcOrchagent invokes SAI APIs to inject the multicast route entry information into ASIC_DB. <br> <br>
+
+7. Syncd being an ASIC_DB subscriber, it receives the multicast route information updated to ASIC_DB by ipmcOrchagent. <br> <br>
+
+8. Syncd processes the multicast route information and invokes the SAI API to inject the (\*, G)  mroute entry into the corresponding ASIC-driver. <br> <br>
+
+9. The multicast route is finally programmed in the hardware.
+
+#### Case 2: Flow for (S,G) route update
+
+1. PIM ASM creates (S,G) entry once the data traffic starts to flow. <br> <br>
+  1.1 When FHR receives the unknown multicast data from a source it traps and delivers the multicast data to the Linux kernel. Kernel delivers this packet to the Pimd process. On reception of data, Pimd starts the registration process. 
+As FHR is not aware of the interested receivers,it encapsulates the data packet in a unicast PIM register message and forwards it to RP address. <br> <br> 
+  1.2  RP's kernel receives the PIM register packet. It decapsulates the data and forwards the data to the downstream routers and it delivers the register packet to Pimd. Pimd processes the register message and it creates (S,G) join by inheriting the OIF from the (\*,G) route along with cpu port and IIF as the incoming interface of the data and  forwards it towards the FHR.
```

Comment:
Please confirm kernel forwards the Reg decap packet or pimd.  if there is detailed design document or any document about FRR behavior, its good idea to have reference in this document.

Reply:
FRR expects kernel to decapsulate and forward the packet. For now, kernel is not forwarding the packet because of a bug. That issue has been fixed in kernel and patch would be applied to achieve this